podpac/core/data/datasource.py

```diff
@@ -475,75 +487,20 @@ def get_native_coordinates(self):
-            return self.native_coordinates
-        else:
-            raise NotImplementedError(
-                "{0}.native_coordinates is not defined and "
-                "{0}.get_native_coordinates() is not implemented".format(self.__class__.__name__)
-            )
+    def set_native_coordinates(self, coordinates, force=False):
+        """ Set the native_coordinates. Used by Compositors as an optimization.
 
-    @property
-    @common_doc(COMMON_DATA_DOC)
-    def base_definition(self):
-        """Base node definition for DataSource nodes.
-        
-        Returns
-        -------
-        {definition_return}
+        Arguments
+        ---------
+        coordinates : :class:`podpac.Coordinates`
+            Coordinates to set. Usually these are coordinates that are shared across compositor sources.
         """
 
-        d = super(DataSource, self).base_definition
-
-        # check attrs and remove unnecesary attrs
-        attrs = d.get("attrs", {})
-        if "source" in attrs:
-            raise NodeException("The 'source' property cannot be tagged as an 'attr'")
-        if "interpolation" in attrs:
-            raise NodeException("The 'interpolation' property cannot be tagged as an 'attr'")
-        if "nan_vals" in attrs and not self.nan_vals:
-            del attrs["nan_vals"]
-
-        # set source or lookup_source
-        if isinstance(self.source, Node):
-            d["lookup_source"] = self.source
-        elif isinstance(self.source, np.ndarray):
-            d["source"] = self.source.tolist()
-        else:
-            d["source"] = self.source
-
-        # assign the interpolation definition
-        d["interpolation"] = self.interpolation
-
-        return d
-
-    # ------------------------------------------------------------------------------------------------------------------
-    # Operators/Magic Methods
-    # ------------------------------------------------------------------------------------------------------------------
-    def __repr__(self):
-        source_name = str(self.__class__.__name__)
-
-        rep = "{}".format(source_name)
-        if source_name != "DataSource":
-            rep += " DataSource"
-
-        source_disp = self.source if isinstance(self.source, string_types) else "\n{}".format(self.source)
-        rep += "\n\tsource: {}".format(source_disp)
-        if trait_is_defined(self, "native_coordinates"):
-            rep += "\n\tnative_coordinates: "
-            for c in self.native_coordinates.values():
-                if isinstance(c, Coordinates1d):
-                    rep += "\n\t\t%s: %s" % (c.name, c)
-                elif isinstance(c, StackedCoordinates):
-                    for _c in c:
-                        rep += "\n\t\t%s[%s]: %s" % (c.name, _c.name, _c)
-
-                # rep += '{}: {}'.format(c.name, c)
-        rep += "\n\tinterpolation: {}".format(self.interpolation)
-
-        return rep
+        if not self.trait_is_defined("_native_coordinates"):
```

Comment:
WOW that's a lot cleaner. Nice. NIIICEE.

Reply:
This went through a silly number of iterations.